crypto/bn/bn_prime.c

```diff
@@ -171,23 +168,41 @@ int BN_generate_prime_ex(BIGNUM *ret, int bits, int safe,
  err:
     OPENSSL_free(mods);
     BN_CTX_end(ctx);
-    BN_CTX_free(ctx);
     bn_check_top(ret);
     return found;
 }
 
+#ifndef FIPS_MODE
+int BN_generate_prime_ex(BIGNUM *ret, int bits, int safe,
+                         const BIGNUM *add, const BIGNUM *rem, BN_GENCB *cb)
+{
+    BN_CTX *ctx = BN_CTX_new();
+    int retval;
+
+    if (ctx == NULL)
+        return 0;
+
+    retval = BN_generate_prime_ex2(ret, bits, safe, add, rem, cb, ctx);
+
+    BN_CTX_free(ctx);
+    return retval;
+}
+#endif
+
 int BN_is_prime_ex(const BIGNUM *a, int checks, BN_CTX *ctx_passed,
                    BN_GENCB *cb)
 {
     return BN_is_prime_fasttest_ex(a, checks, ctx_passed, 0, cb);
 }
 
 /* See FIPS 186-4 C.3.1 Miller Rabin Probabilistic Primality Test. */
-int BN_is_prime_fasttest_ex(const BIGNUM *w, int checks, BN_CTX *ctx_passed,
+int BN_is_prime_fasttest_ex(const BIGNUM *w, int checks, BN_CTX *ctx,
                             int do_trial_division, BN_GENCB *cb)
 {
     int i, status, ret = -1;
-    BN_CTX *ctx = NULL;
+#ifndef FIPS_MODE
```

Comment:
OK this is kind of scaring me.
Why do you allow this in non FIPS?
I have a concern that this is going to bite us at some point - especially since we are not running the 
tests using the FIPS provider yet..

Reply:
Yes, at least there is a missing if (ctx == NULL) goto err; in FIPS_MODE

Reply:
I'm not sure I understand why it's disabled for FIPS. Is there some weird consensus that BN_CTX shouldn't be used in the FIPS module?

Reply:
The BN_CTX needs to be passed because the FIPS module needs the non-default OPENSSL_CTX (which is inside the passed-in BN_CTX). The goal is quite clear.

What is not clear to me (but that might be because I did not study the provider implementation plan well enough) how this will all work from the application side - if the application calls BN_generate_prime_ex2 by itself what will happen? As I understand it now the default implementation be called even with FIPS provider being loaded?

Reply:
If the application calls BN code itself, then it's getting to the libcrypto build, not the FIPS build.

Reply:
Then it would still be beneficial to somehow guard against misuse of these APIs inside the FIPS provider itself. Coverity scans will hopefully reveal this but could we make it more explicit such as with __attribute__((nonnull)) annotations? However they would have to be FIPS_MODE only.

Reply:
Not sure why this is considered scary. I could put in the `if (ctx == NULL) goto err;` line for FIPS_MODE if that lessens your worries?

> Why do you allow this in non FIPS?

Outside of FIPS_MODE the presence of a `BN_CTX` is simply an optimisation. It allows the reuse of temporary BIGNUM objects. Inside FIPS_MODE it must be present because we get the current OPENSSL_CTX from it. Since FIPS_MODE requires an explicit OPENSSL_CTX (the default OPENSSL_CTX is not valid within the FIPS provider), we *must* always have a `BN_CTX`.